util/pyclient/bft_client.py

```diff
@@ -428,24 +468,25 @@ async def _receive_from_replica(self, dest_addr, replicas_addr, required_replies
         3) Process the received message payload.
         If there is an error in stages 1 or 2 - exit straight (connection is closed inside _stream_recv_some)
         """
-        if dest_addr not in self.ssl_streams:
-            self.establish_ssl_stream_parklot[dest_addr].unpark()
-            return
-        data = bytearray()
-        stream = self.ssl_streams[dest_addr]
-        while len(data) < self.MSG_HEADER_SIZE:
-            if not await self._stream_recv_some(data, dest_addr, stream, self.MSG_HEADER_SIZE):
+        while True:
```

Comment:
why is this additional while loop needed?

Reply:
The _recv_data method of the TCP client implementation is supposed to receive data and call _process_received_msg until the the enclosing cancel_scope times out just like the UDP implementation. Instead it would return. With normal messages with small replies it wasn't obvious, but with batches that wait for many replies it fails to reach quorum, because when it returns it's as if the retry timeout has been reached.

Reply:
Thanks for clarifying.